### PR TITLE
nixos: Add option to influence override strategies in systemd units

### PR DESCRIPTION
**Pull request by ck3d**
###### Description of changes

This PR add a new unit option named systemdRepresentation and one example how it can be used.
You can control the represenation of configuration information for systemd, which is mandatory for template units.
This option can be used to fix following issues:

- #80933
- #135557

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by aanderse** [comment]
Cool idea. After a quick look the implementation looks fine to me. Maybe a nitpick but can we change the option name? `systemdRepresentation` is a bit of a mouthful and the `systemd` part seems redundant.

My suggestion for a better name?

🤷‍♂️

**Comment by ck3d** [comment]
Thanks for the review. I am also fine without the `systemd` prefix.

**Comment by markuskowa** [comment]
> My suggestion for a better name?

Here are some alternative, shorter suggestions:
* `nix2systemd`
* `unitMapping`
* `creationMode`

**Comment by aanderse** [comment]
> > My suggestion for a better name?
> 
> Here are some alternative, shorter suggestions:
> 
>     * `nix2systemd`
> 
>     * `unitMapping`
> 
>     * `creationMode`

`creationMode` has a certain ring to it... but let's here what others have to say too :smile:

File: nixos/lib/systemd-unit-options.nix [hunk 1]
```diff
@@ -48,6 +48,18 @@ in rec {
...
     };
 
+    representation = mkOption {
+      default = "auto";
+      type = types.enum [ "auto" "drop-in" ];
+      description = lib.mdDoc ''
+        Defines how unit configuration is represented for systemd:
+        `auto` creates a unit file when no unit file is available
+        otherwise a drop-in.
+        `drop-in` creates a drop-in file named `overrides.conf`.
```
Comment (claes) [inline nixos/lib/systemd-unit-options.nix:58]: Perhaps add something like 
"Use 'drop-in' for template units (units with @ in the name)" 
to make it more clear to the user when this is intended to be used?

File: nixos/lib/systemd-unit-options.nix [hunk 2]
```diff
@@ -48,6 +48,18 @@ in rec {
       '';
     };
 
+    representation = mkOption {
```
Comment (flokli) [inline nixos/lib/systemd-unit-options.nix:51]: What about `useDropin` or `asDropin` instead? I'd probably not understand what this is doing, just by looking at the configuration name.

Comment (ck3d) [reply nixos/lib/systemd-unit-options.nix:51]: Thanks, I used now asDropin and also renamed auto

File: nixos/lib/systemd-unit-options.nix (hunk 1 above)
Comment (ck3d) [reply nixos/lib/systemd-unit-options.nix]: Good point, I added an example

**Comment by ck3d** [comment]
I talked with @Mic92 at NixCon 2022 an alternative to `representation` which is `overrideStrategy`.

**Comment by aanderse** [comment]
> I talked with @Mic92 at NixCon 2022 an alternative to `representation` which is `overrideStrategy`.

Yeah, `overrideStrategy` sounds nice too.

**Comment by ck3d** [comment]
@ofborg test systemd-machinectl

**Comment by ck3d** [comment]
I rebased the PR and renamed `representation` to `overrideStrategie`.

File: nixos/lib/systemd-lib.nix [hunk 3]
```diff
@@ -187,11 +187,14 @@ in rec {
         done
       done
 
-      # Symlink all units defined by systemd.units. If these are also
-      # provided by systemd or systemd.packages, then add them as
+      # Symlink units defined by systemd.units where override strategie
```
[the remark this answers is not in the document]
Comment (ck3d) [reply nixos/lib/systemd-lib.nix]: Thanks, updated my PR

**Comment by flokli** [comment]
The test now fails to build:

```
❯ nix-build nixos/tests/systemd-machinectl.nix
error: attribute 'overrideStrategy' missing

       at /home/flokli/dev/nixos/nixpkgs/nixos/lib/systemd-lib.nix:197:35:

          196|           (n: v: v.unit)
          197|           (lib.filterAttrs (n: v: v.overrideStrategy == "asDropinIfExists") units))}; do
             |                                   ^
          198|         fn=$(basename $i/*)
(use '--show-trace' to show detailed location information)

```

**Comment by ck3d** [comment]
Thanks for the catch. The rebase introduced my own PR #189703 . I should have seen this issue coming. I updated this PR with a fix.

**Comment by ck3d** [comment]
@ofborg test systemd-machinectl

**Comment by flokli** [comment]
@ck3d can you check back on https://github.com/NixOS/nixpkgs/issues/80933 and https://github.com/NixOS/nixpkgs/issues/135557 ? Maybe we need some more prominent documentation of this feature, and/or a release note entry.

**Comment by ck3d** [comment]
Good point, I commented and closed both issues and created a PR to add release note for new option overrideStrategy, see
https://github.com/NixOS/nixpkgs/pull/198336

**Comment by bamhm182** [comment]
I'm having issues getting this to work, but it seems like it should be working if I am understanding it properly.

I have the following code within [this file](https://gitlab.com/bytepen/nixos/public/-/blob/wip-podman/src/common/services/podman.nix):
```
systemd.services = {
      ...
      "podman-compose@" = {
        enable = true;
        after = [ "podman-sys-prep.service" ];
        description = "Systemd service to manage podman-compose for %i";
        ...
        serviceConfig = {
          ...
          WorkingDirectory = "/etc/containers/compose/%i";
          ...
        };
      };
    };
```

and the following code within [this file](https://gitlab.com/bytepen/nixos/public/-/blob/wip-podman/src/common/containers/nextcloud/default.nix):
```
    systemd.services."podman-compose@nextcloud" = {
      wantedBy = [ "machines.target" ];
      overrideStrategy = "asDropinIfExists";
    };
```

However, that code doesn't link podman-compose@nextcloud.service to podman-compose@.service like I feel like this PR is stating it should. Instead, it creates a largely empty podman-compose@nextcloud.service which fails due to not having any of the required Exec lines. I have tried asDropin and asDropinIfExist, but neither work right for me.

**Comment by ck3d** [comment]
The issue you face is probably the same as #219013 .

**Pull request merged**